openeo-geotrellis/src/main/scala/org/openeo/geotrellis/package.scala

```diff
@@ -158,28 +160,67 @@ package object geotrellis {
-        action
-      }
+
+  /**
+   * taken from DefaultProcessApi
+   */
+  def withRetryAfterRetries[R](context: String)(fn: => HttpResponse[R])(implicit logger: Logger): HttpResponse[R] = {
+    var lastResponse: Option[HttpResponse[R]] = None
+
+    val retryable: Throwable => Boolean = {
+      case HttpStatusException(400, _, responseBody) if responseBody.contains("Request body should be non-empty.") => true
+      case HttpStatusException(statusCode, _, responseBody) if statusCode >= 500
+        && !responseBody.contains("newLimit > capacity") && !responseBody.contains("Illegal request to https") => true
+      case _: SocketTimeoutException => true
+      case _: CirceException => true
+      case e => logger.error(s"Not attempting to retry unrecoverable error in context: $context", e); false
     }
 
-    attempt(retries = 4, delay = (30, SECONDS)) {
-      action
+    val shakyConnectionRetryPolicy = new net.jodah.failsafe.RetryPolicy[HttpResponse[R]]()
+      .handleIf(retryable.asJava)
+      .withBackoff(1, 1000, ChronoUnit.SECONDS) // should not reach maxDelay because of maxAttempts 5
+      .withJitter(0.5)
+      .withMaxAttempts(5)
+      .onFailedAttempt((attempt: ExecutionAttemptedEvent[HttpResponse[R]]) => {
+        val e = attempt.getLastFailure
+        logger.warn(s"Attempt ${attempt.getAttemptCount} failed in context: $context", e)
+      })
+      .onFailure((execution: ExecutionCompletedEvent[HttpResponse[R]]) => {
+        val e = execution.getFailure
+        logger.error(s"Failed after ${execution.getAttemptCount} attempt(s) in context: $context", e)
+      })
+
+    val isRateLimitingResponse: Throwable => Boolean = {
+      case HttpStatusException(429, _, _) => true
+      case _ => false
     }
+
+    var retryAfterSecondsCounter = 20.0
+    val rateLimitingRetryPolicy = new net.jodah.failsafe.RetryPolicy[HttpResponse[R]]()
+      .handleIf(isRateLimitingResponse.asJava)
+      .withMaxAttempts(5)
+      .withDelay((_: HttpResponse[R], _: HttpStatusException, _: ExecutionContext) => {
+        val retryAfterSeconds = lastResponse match {
+          case None => retryAfterSecondsCounter.toLong
+          case Some(x) => x
+            .headers
+            .find { case (header, _) => header equalsIgnoreCase "retry-after" }
+            .map { case (_, values) => values.head.toLong }
+            .getOrElse(retryAfterSecondsCounter.toLong)
+        }
+        retryAfterSecondsCounter *= 1.6
+        Duration.ofSeconds(retryAfterSeconds)
+      })
+      .onRetryScheduled((retry: ExecutionScheduledEvent[scalaj.http.HttpResponse[_root_.geotrellis.raster.io.geotiff.MultibandGeoTiff]]) => {
+        logger.warn(s"Scheduled retry within ${retry.getDelay} because of 429 response in context: $context")
+      })
+
+    Failsafe
+      .`with`(java.util.Arrays.asList(shakyConnectionRetryPolicy, rateLimitingRetryPolicy))
+      .get(() => {
+        val res = fn
+        lastResponse = Some(res)
+        res.throwError
```

Comment:
```suggestion
.get(() => httpResponse.throwError)
```

Reply:
Refactored to not be dependant on `throwError`

Reply:
With the `throwError` flow it is hacky to access the `Retry-After` header. Ok to only keep the without throwError flow?

Reply:
Quite right, you don't have the response headers in that case. :thinking:

Reply:
There was no real need to support the .throwError logic. Support dropped and duplicate code cleaned up!